### PR TITLE
Fix integer overflow in range

### LINKED ISSUE #64081: slient overflow occurs in tf.range leading to incorrect result, here is a possible fix
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.15.0

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

Following the documentation https://www.tensorflow.org/api_docs/python/tf/range, tf.range should have consistent result with np.arange. 
However, given the following inputs, tf.range outputs `tf.Tensor([1136033460], shape=(1,), dtype=int32)`, which is **incorrect**
The correct result is np.arange's output `[1136033460 -713794229]`.

I did some analysis and find the root cause lies in here:
https://github.com/tensorflow/tensorflow/blob/06e37b8d7f2ba0338efd0f588f068122414f0722/tensorflow/core/kernels/sequence_ops.cc#L100

When computing the `limit - start`, the actual value overflows.
To solve this problem, we can change the original code to a new one by applying the division first
origin:
```
Eigen::numext::ceil(Eigen::numext::abs((limit - start) / delta));
```

new:
```
Eigen::numext::ceil(Eigen::numext::abs(limit / delta - start / delta);
```

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf
start = 1136033460
end = -2110457150
step = -1849827689
out = tf.range(start, end, step)
print(out)
out_np = np.arange(start, end, step)
print(out_np)
```
```


### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by mattbahr**
Fixes #64081

I modified the fixes from the rolled back PR's #77018 and #76252 to resolve an overflow that was identified by one of the new test cases. The issue was that in `ragged_range_op.cc`, `value` is incremented at the end of the loop logic, which results in a needless addition on the final iteration and creates the possibility for an overflow. This had no impact on the output of the operation and did not cause any of the tests to fail, but it was picked up by ASan.

https://github.com/tensorflow/tensorflow/blob/ec9e1531f0114662baef0f3985de5620b43d0d29/tensorflow/core/kernels/ragged_range_op.cc#L123-L126

I added a condition that only increments the value if there are more iterations to be completed.

### CODE REVIEW

**Review by mihaimaruseac — COMMENTED** [review]
I think this LGTMs, but let's also see if @cantonios sees any downsides on this patch

**Comment by mattbahr** [comment]
@cantonios When you get the chance, would you mind taking a look at this review? Thanks!

File: tensorflow/core/kernels/sequence_ops.cc [hunk 1]
```diff
@@ -93,8 +95,21 @@ class RangeOp : public OpKernel {
...
+      } else {
+        range = static_cast<uint64_t>(Eigen::numext::abs(limit - start));
+      }
+
+      uint64_t size_unsigned = Eigen::divup(range,
+                    static_cast<uint64_t>(Eigen::numext::abs(delta)));
+      OP_REQUIRES(context,
+                  size_unsigned <= std::numeric_limits<int64_t>::max(),
+                  InvalidArgument("Requires ((limit - start) / delta) <= ",
+                                  std::numeric_limits<int64_t>::max()));
```
Comment (cantonios) [inline tensorflow/core/kernels/sequence_ops.cc:111]: Does it _actually_ need to be less than int64_t max, or could it be uint64_t max (i.e. no real limit)?

**Review by cantonios — CHANGES_REQUESTED** [review]
See question about limit.

**Comment by mattbahr** [comment]
@cantonios Leaving the size unsigned can cause a signed integer overflow when it is passed to the AddDimWithStatus function.

https://github.com/tensorflow/tensorflow/blob/137b1e0e6890e8bad125911a325846db4ab05516/tensorflow/core/kernels/sequence_ops.cc#L109

This function expects type int64_t, and for a very large size will interpret it as negative.

https://github.com/tensorflow/tensorflow/blob/137b1e0e6890e8bad125911a325846db4ab05516/tensorflow/core/framework/tensor_shape.cc#L427-L433

**Review by cantonios — APPROVED** [review]

**Pull request merged**